Sources/NIOCore/ChannelPipeline.swift

```diff
@@ -2089,3 +2091,171 @@ extension ChannelPipeline: CustomDebugStringConvertible {
+    ///
+    /// - parameters:
+    ///     - handler: the channel handler object whose outbound buffered bytes will be audited.
+    public func auditOutboundBufferedBytes(handler: ChannelHandler) -> EventLoopFuture<Int> {
+        let future: EventLoopFuture<Int>
+        
+        if self.eventLoop.inEventLoop {
+            future = self.eventLoop.makeCompletedFuture(audit0(handler: handler, direction: .outbound))
+        } else {
+            future = self.eventLoop.submit {
+                try self.audit0(handler: handler, direction: .outbound).get()
+            }
+        }
+        
+        return future
+    }
+    
+    /// Audit the total numbers of inbound bytes buffered in the channel handler with the given `name`.
+    ///
+    /// - parameters:
+    ///     - name: the name of the channel handler whose inbound buffered bytes will be audited.
+    public func auditInboundBufferedBytes(name: String) -> EventLoopFuture<Int> {
+        let future: EventLoopFuture<Int>
+        
+        if self.eventLoop.inEventLoop {
+            future = self.eventLoop.makeCompletedFuture(audit0(name: name, direction: .inbound))
+        } else {
+            future = self.eventLoop.submit {
+                try self.audit0(name: name, direction: .inbound).get()
+            }
+        }
+        
+        return future
+    }
+    
+    /// Audit the total numbers of inbound bytes buffered in the channel handler with the given `handler`.
+    ///
+    /// - parameters:
+    ///     - handler: the channel handler object whose inbound buffered bytes will be audited.
+    public func auditInboundBufferedBytes(handler: ChannelHandler) -> EventLoopFuture<Int> {
+        let future: EventLoopFuture<Int>
+        
+        if self.eventLoop.inEventLoop {
+            future = self.eventLoop.makeCompletedFuture(audit0(handler: handler, direction: .inbound))
+        } else {
+            future = self.eventLoop.submit {
+                try self.audit0(handler: handler, direction: .inbound).get()
+            }
+        }
+        
+        return future
+    }
+    
+    private func audit0(name: String, direction: AuditDirection) -> Result<Int, Error> {
+        let result = self.contextSync(name: name)
+        switch result {
+        case .success(let context):
+            return audit0(context: context, direction: direction)
+        case .failure(let error):
+            return .failure(error)
+        }
+    }
+    
+    private func audit0(handler: ChannelHandler, direction: AuditDirection) -> Result<Int, Error> {
+        let result = self.contextSync(handler: handler)
+        switch result {
+        case .success(let context):
+            return audit0(context: context, direction: direction)
+        case .failure(let error):
+            return .failure(error)
+        }
+    }
+    
+    private func audit0(context: ChannelHandlerContext, direction: AuditDirection) -> Result<Int, Error> {
```

Comment:
I'm not sure this needs an error result for not being able to do the type conversion, it might just want to return `Int?`

Reply:
It could just throw if it really wants to propagate an error.

Reply:
Given that it is not a good design choice to add an error in `ChannelPipelineError` and error conveys the same outcome as `Int?`, I agree it is better to return `Int?` directly.